### ISSUE TITLE
Assert failed with two functions, an optional, and typeInfo

### ISSUE THREAD

**Issue by SpexGuy** [ziglang/zig#6942]
I wasn't able to get the repro any smaller than this.  It seems like all of this is necessary to trip the assert.
```zig
fn optional(comptime PtrOrSliceT: type) ?u29 {
    // replacing this with a constant of type u29 doesn't crash.
    // so it seems like type info is important
    return @typeInfo(PtrOrSliceT).Pointer.alignment;
    // but this works for some reason:
    //return if (false) void else @as(?u29, @typeInfo(PtrOrSliceT).Pointer.alignment);
}

fn sink(comptime alignment: ?u29) void { }

test "mem.Allocator basics" {
    // this compiles fine
    //const x = comptime optional([]u8);
    // so does this
    //sink(@typeInfo([]u8).Pointer.alignment);
    // both functions are necessary for some reason
    sink(comptime optional([]u8));
}
```
```batch
> zig version
0.6.0+0088efc4b
> zig test test.zig
Assertion failed at D:\a\1\s\src\stage1\analyze.cpp:5481 in hash_const_val. This is a bug in the Zig compiler.
Unable to dump stack trace: debug info stripped
```

**Comment by LemonBoy** [comment]
Duplicate of #7108, same root cause but trips a different assertion.

**Comment by mikdusan** [comment]
with latest compiler:
- compiler no longer asserts and test passes
- appears this was fixed between 0.8.1 and 0.9.0

<details>
<summary>repro.zig</summary>

```zig
fn optional(comptime PtrOrSliceT: type) ?u29 {
    // replacing this with a constant of type u29 doesn't crash.
    // so it seems like type info is important
    return @typeInfo(PtrOrSliceT).Pointer.alignment;
    // but this works for some reason:
    //return if (false) void else @as(?u29, @typeInfo(PtrOrSliceT).Pointer.alignment);
}

fn sink(comptime _: ?u29) void {}

test "mem.Allocator basics" {
    // this compiles fine
    //const x = comptime optional([]u8);
    // so does this
    //sink(@typeInfo([]u8).Pointer.alignment);
    // both functions are necessary for some reason
    sink(comptime optional([]u8));
}
```
</details>

```sh
$ uname -a
Linux archlinux 6.1.2-arch1-1 #1 SMP PREEMPT_DYNAMIC Sat, 31 Dec 2022 17:40:35 +0000 x86_64 GNU/Linux

$ zig version
0.11.0-dev.1247+87b223428

$ zig test repro.zig |& cat
1/1 test.mem.Allocator basics... OK
All 1 tests passed.
```
